File: Java/777.java
__________________________________________________________________________________________________
sample 1 ms submission
class Solution {
    public boolean canTransform(String start, String end) {
        // https://www.cnblogs.com/grandyang/p/9001474.html
        char[] arr1 = start.toCharArray();
        char[] arr2 = end.toCharArray();
        int len = arr1.length, i = 0, j = 0;
        while(i < len && j < len) {
            while(i < len && arr1[i] == 'X')   i++;
            while(j < len && arr2[j] == 'X')   j++;
            if((i < len) ^ (j < len))   return false;
            if(i < len && j < len) {
                if(arr1[i] != arr2[j])  return false;
                if(arr1[i] == 'L' && i < j) return false;
                if(arr1[i] == 'R' && i > j) return false;
                i++;
                j++;
            }
        }
        return true;
    }
}
__________________________________________________________________________________________________
sample 37240 kb submission
class Solution {
    public boolean canTransform(String start, String end) {
        int i = 0, j = 0;
        int length = start.length();
        
        while(i < length && j < length)
        {
            while(i < length && start.charAt(i) == 'X') i++;
            while(j < length && end.charAt(j) == 'X') j++;
            if((i == length) && (j == length))
                return true;
            if((i < length) ^ (j < length))
                return false;
            char c1 = start.charAt(i);
            char c2 = end.charAt(j);
            if((c1 != c2) || (c1 == 'R') && (i>j) || (c2 == 'L') && (i<j))
                return false;
            i++;
            j++;
        }
        return true;
    }
}
__________________________________________________________________________________________________

File: Python3/777.py
__________________________________________________________________________________________________
sample 28 ms submission
class Solution:
    def canTransform(self, s: str, t: str) -> bool:
        if not s or not t:
            return not s and not t
        if len(s) != len(t):
            return False
        
        n = len(s)
        i = j = 0
        
        while i < n or j < n:
            while i < n and s[i] == 'X':
                i += 1
            while j < n and t[j] == 'X':
                j += 1
            
            if i >= n or j >= n:
                return i >= n and j >= n
            
            if s[i] != t[j]: return False
            if s[i] == 'L' and i < j: return False
            if s[i] == 'R' and i > j: return False
            
            i += 1
            j += 1
        
        return i >= n and j >= n
__________________________________________________________________________________________________
sample 13172 kb submission
class Solution:
    def canTransform(self, start: str, end: str) -> bool:
        if start.replace('X', '') != end.replace('X', ''):
            return False
        t = 0
        for i in range(0, len(start)):
            if start[i] == 'L':
                while t < len(end) and end[t] != 'L':
                    t += 1
                if i < t:return False
                t += 1
        t = 0
        for i in range(0, len(start)):
            if start[i] == 'R':
                while t < len(end) and end[t] != 'R':
                    t += 1
                if i > t:return False
                t += 1
        return True
__________________________________________________________________________________________________
